src/main.rs

```diff
@@ -782,6 +796,85 @@ fn start_vmm(cmd_arguments: ArgMatches) -> Result<Option<String>, Error> {
     r.map(|_| api_socket_path)
 }
 
+// This is a best-effort solution to the latency induced by the RCU
+// synchronization that happens in the kernel whenever the file
+// descriptor table is full and needs to be expanded. The initial table
+// has 64 entries on amd64.
+// Rust programs that use {File,EventFd}::try_clone() to share handles
+// are impacted by this issue.
+// This behavior is quite noticeable in the restore snapshot scenario,
+// the latency is a big chunk of the total time required to start
+// cloud-hypervisor and to restore the snapshot.
+// The kernel has an optimization in the expand table code path where it
+// doesn't call synchronize_rcu() if there is only one thread in the
+// process.
+// To address this issue we attempt to expand the descriptor table at
+// application start, when it has only one thread.
+fn expand_fdtable() -> Result<(), FdTableError> {
+    let mut limits = libc::rlimit {
+        rlim_cur: 0,
+        rlim_max: 0,
+    };
+
+    // SAFETY: FFI call with valid arguments
+    if unsafe { libc::getrlimit(libc::RLIMIT_NOFILE, &mut limits) } < 0 {
+        return Err(FdTableError::GetRLimit(io::Error::last_os_error()));
+    }
+
+    let dummy_evt = EventFd::new(EFD_NONBLOCK).map_err(FdTableError::CreateEventFd)?;
+
+    let mut table_size: i32;
+
+    if limits.rlim_cur < 4096 {
+        table_size = limits.rlim_cur as i32;
+
+        let new_limits = libc::rlimit {
+            rlim_cur: std::cmp::min(4096, limits.rlim_max),
+            rlim_max: limits.rlim_max,
+        };
+
+        // If this fails we are using the current rlim_cur which is lower
+        // than 4096.
+        // SAFETY: FFI call with valid arguments
+        if unsafe { libc::setrlimit(libc::RLIMIT_NOFILE, &new_limits) } == 0 {
+            table_size = new_limits.rlim_cur as i32;
+        }
+    } else {
+        table_size = 4096;
+    }
```

Comment:
I wonder why not adopt the simpler behavior  as done from firecracker, e.g. expand to match RLIMIT_NOFILE or a default value (say 4096) if no RLIMIT_NOFILE is set? I think it is okay that we don't set `setrlimit` even if the default value (say 4096) is larger than the hard limit.

Reply:
There are a few cases where the implementation from firecracker falls short, for example:

1. rlim_cur = 64, rlim_max = infinity. The table will be expanded to 64 entries, which doesn't fix the issue.
2. rlim_cur = 4M, rlim_max = infinity. The table will occopy 4M * sizeof(void*), 32MB per process on amd64. This leads to unnecesary memory usage.